tests/functional/test_download.py

```diff
@@ -0,0 +1,315 @@
+        future = self.manager.download(**call_kwargs)
+
+        with self.assertRaises(IOError):
+            future.result()
+
+    def test_retries_and_succeeds(self):
+        stubbed_responses = self.stubbed_responses
+        # Insert a response that will trigger a retry.
+        stubbed_responses.insert(
+            1,
+            {
+                'method': 'get_object',
+                'service_response': {
+                    'Body': StreamWithError(SOCKET_ERROR)
+                }
+            }
+        )
+
+        for stubbed_response in stubbed_responses:
+            self.stubber.add_response(**stubbed_response)
+
+        future = self.manager.download(**self.call_kwargs)
+        future.result()
+
+        # The retry should have been consumed and the process should have
+        # continued using the successful responses.
+        self.stubber.assert_no_pending_responses()
+        with open(self.filename, 'rb') as f:
+            self.assertEqual(self.content, f.read())
+
+    def test_retry_failure(self):
+        stubbed_responses = []
+        # Add the HeadObject to the stubbed responses
+        stubbed_responses.append(self.stubbed_responses[0])
+
+        # Make responses that trigger three retries
+        max_retries = 3
+        self.config.num_download_attempts = max_retries
+        self._manager = TransferManager(self.client, self.config)
+        for _ in range(max_retries):
+            stubbed_responses.append(
+                {
+                    'method': 'get_object',
+                    'service_response': {
+                        'Body': StreamWithError(SOCKET_ERROR)
+                    }
+                }
+            )
+
+        for stubbed_response in stubbed_responses:
+            self.stubber.add_response(**stubbed_response)
+
+        future = self.manager.download(**self.call_kwargs)
+
+        # A retry exceeded error should have happened.
+        with self.assertRaises(RetriesExceededError):
+            future.result()
+
+        # All of the retries should have been used up.
+        self.stubber.assert_no_pending_responses()
+
+    def test_can_provide_file_size(self):
+        stubbed_responses = self.stubbed_responses
+        # Remove the HeadObject response
+        stubbed_responses.pop(0)
+
+        for stubbed_response in stubbed_responses:
+            self.stubber.add_response(**stubbed_response)
+
+        call_kwargs = self.call_kwargs
+        call_kwargs['subscribers'] = [FileSizeProvider(len(self.content))]
+
+        future = self.manager.download(**call_kwargs)
+        future.result()
+
+        # The HeadObject should have not happened and should have been able
+        # to successfully download the file.
+        self.stubber.assert_no_pending_responses()
+        with open(self.filename, 'rb') as f:
+            self.assertEqual(self.content, f.read())
+
+
+class TestNonRangedDownload(BaseDownloadTest):
+    __test__ = True
```

Comment:
I have not found a better way. I am up for any suggestions. I also considered adding skips but seemed undesirable as it would result in a bunch of skips in the test output.